api/TODO.adoc

```diff
@@ -0,0 +1,119 @@
+== Naming conventions
+
+Should we use plain names for classes/interfaces, using the package name to namespace them, or should we prefix them.
+
+`org.infinispan.api.sync.Cache`
+
+vs
+
+`org.infinispan.api.sync.SyncCache`
+
+== API dependencies
+
+The sync and async API must depend only on core JDK classes.
+The mutiny API has an optional dependency on mutiny.
+
+Should we depend on `javax.transaction` for the Transaction class or should we provide our own interface and a way to
+convert ?
+
+== API traversal
+
+The API is structured like a tree as follows:
+
+ infinispan->[sync,async,mutiny]->[caches,...]->[cache,...]
+      ^                  ^               ^            ^
+  container           variant         groups      construct
+
+Most of the constructs exist in multiple variants
+
+* do we want to be able to "jump" between different variants of the same construct ? `cache.mutiny()`
+* do we allow "climbing up" the API hierarchy step-by-step ? `cache.caches().container()`
+* do we allow "jumping up" the API hierarchy  ? `cache.container()`
+
+== Listeners
+
+Handle sync listeners with callbacks and async/mutiny listeners with subscriptions.
+
+=== Sync API
+
+==== Single instance callback
+
+Event depends on the interface implemented by the listener instance. Using functional interfaces allows simple lambda
+callbacks:
+
+```java
+cache.listen((KeyValueCreatedListener<String, String>) event -> {/* Handler */});
+```
+
+For handling multiple events with a single instance, implement multiple interfaces:
+
+```java
+public static class AListener implements KeyValueUpdatedListener, KeyValueRemovedListener {
+  @Override
+  public void onRemove(KeyValueEvent event) {/* Handler */}
+
+  @Override
+  public void onUpdate(KeyValueEvent event) {/* Handler */}
+}
+```
+
+==== Builder-style
+
+```java
+cache.listen()
+    .includeCurrentState()
+    .observation(Observation.POST)
+    .onCreate(event -> {/* Handler */});
+```
+
+==== Listener removal
+
+Registering a listener in the sync API returns a handle to the listener which can be used for removal:
+
+```java
+Handle<KeyValueListener> handle = cache.listen(mylistener);
```

Comment:
can we use `AutoCloseable` instead of creating a new interface `Handle`?

Reply:
Good call !

Reply:
One slight issue with `AutoCloseable` is the `close()` signature throws `Exception` which is a bit annoying.

Reply:
I was originally  thinking of using AutoCloseable but I liked being able to retrieve the listener later if needed as well. Also if we keep Handle we can implement AutoCloseable but override close to not throw an exception btw.

Reply:
Another little snag: I'm sure you want the async version to be `CompletionStage<Void> close()`, so `AutoCloseable` is a no go in non-sync scenarios.

Reply:
While answering a comment above, I noticed that in Reactive Streams, `Subscription#cancel()` is void, so I'm a bit unsure.

Reply:
we can extend `AutoCloseable` and add `asyncClose()` for async
Same thing for `Subscription` and add `asyncCancel()`. 

I guess reactive-streams wasn't designed with distributed systems in mind :)

Reply:
That is an interesting point about stop. I can get behind keeping it sync and it just removes the listener from the local node and then sends the removal asynchronously from the other nodes. But we need to for sure log any issues as ERROR then as this will not be noticed by our tests even.

Reply:
> One slight issue with `AutoCloseable` is the `close()` signature throws `Exception` which is a bit annoying.

This can be overloaded in the Handle interface.  That way one can use try-with-resource semantics without having to handle the exception.

> I'm not sure `asyncClose()` is necessary. 

👍

Reply:
> But we need to for sure log any issues as ERROR then as this will not be noticed by our tests even.

It would definitely require some work to test that all resources associated with the listener are removed, but it should still be possible.
We could also add a `closeAndReport()` method that returns a CompletionStage which reports any errors, but it looks a bit out of place.

Reply:
I've made `ListenerHandle<T> extends AutoCloseable` and override the `close()` method so it doesn't throw an exception